src/issuance.rs

```diff
@@ -365,44 +434,68 @@ impl IssueBundle<Signed> {
+///
+/// # Errors
+///
+/// * `IssueBundleInvalidSignature`: This error occurs if the signature verification
+///    for the provided `sighash` fails.
+/// * `WrongAssetDescSize`: This error is raised if the asset description size for any
+///    asset in the bundle is incorrect.
+/// * `IssueActionPreviouslyFinalizedAssetBase`:  This error occurs if the asset has already been
+///    finalized (inserted into the `finalized` collection).
+/// * `IssueActionIncorrectAssetBase`: This error occurs if any note has an incorrect note type.
+/// * `ValueSumOverflow`: This error occurs if an overflow happens during the calculation of
+///     the value sum for the notes in the asset.
+/// * `IssueBundleIkMismatchAssetBase`: This error is raised if the `AssetBase` derived from
+///    the `ik` (Issuance Validating Key) and the `asset_desc` (Asset Description) does not match
+///    the expected `AssetBase`.
 pub fn verify_issue_bundle(
     bundle: &IssueBundle<Signed>,
     sighash: [u8; 32],
     finalized: &mut HashSet<AssetBase>, // The finalization set.
-) -> Result<(), Error> {
+) -> Result<SupplyInfo, Error> {
     if let Err(e) = bundle.ik.verify(&sighash, &bundle.authorization.signature) {
         return Err(IssueBundleInvalidSignature(e));
     };
 
-    let s = &mut HashSet::<AssetBase>::new();
-
-    let newly_finalized = bundle
-        .actions()
-        .iter()
-        .try_fold(s, |newly_finalized, action| {
-            if !is_asset_desc_of_valid_size(action.asset_desc()) {
-                return Err(WrongAssetDescSize);
-            }
-
-            // Fail if any note in the IssueAction has incorrect note type.
-            let asset = action.are_note_asset_ids_derived_correctly(bundle.ik())?;
-
-            // Fail if the asset was previously finalized.
-            if finalized.contains(&asset) || newly_finalized.contains(&asset) {
-                return Err(IssueActionPreviouslyFinalizedNoteType(asset));
-            }
-
-            // Add to the finalization set, if needed.
-            if action.is_finalized() {
-                newly_finalized.insert(asset);
-            }
-
-            // Proceed with the new finalization set.
-            Ok(newly_finalized)
-        })?;
+    let supply_info =
+        bundle
+            .actions()
+            .iter()
+            .try_fold(SupplyInfo::default(), |mut supply_info, action| {
+                if !is_asset_desc_of_valid_size(action.asset_desc()) {
+                    return Err(WrongAssetDescSize);
+                }
+
+                let (asset, supply) = action.compute_new_supply(bundle.ik())?;
+
+                // Insert new supply into the supply info (fails if the asset was
+                // previously finalized or if the supply for the asset was already
+                // inserted and finalized)
+                if finalized.contains(&asset)
+                    || (!supply_info.try_insert_asset_supply(asset, supply))
```

Comment:
as discussed above, `add_supply()` should not fail.

The rational is that the consensus actually finalizes an assetBase at the end of each Block, not mid-block nor mid-transaction.

Reply:
Done.